gdnative/src/class.rs

```diff
@@ -382,3 +384,181 @@ godot_class! {
         }
     }
 }
+
+// Class builder
+
+pub type GodotScriptMethodType = unsafe extern "C" fn(
+    *mut sys::godot_object,
+    *mut libc::c_void,
+    *mut libc::c_void,
+    libc::c_int,
+    *mut *mut sys::godot_variant
+) -> sys::godot_variant;
+
+pub type GodotScriptConstructorType = unsafe extern "C" fn(
+    *mut sys::godot_object,
+    *mut libc::c_void
+) -> *mut libc::c_void;
+
+pub type GodotScriptDestructorType = unsafe extern "C" fn(
+    *mut sys::godot_object,
+    *mut libc::c_void,
+    *mut libc::c_void
+) -> ();
+
+pub enum GodotRpcMode {
+    Disabled,
+    Remote,
+    Sync,
+    Mater,
+    Slave
+}
+
+pub struct GodotScriptMethodAttributes {
+    pub rpc_mode: GodotRpcMode
+}
+
+pub struct GodotScriptMethod<'l> {
+    pub name: &'l str,
+    pub method_ptr: Option<GodotScriptMethodType>,
+    pub attributes: GodotScriptMethodAttributes,
+
+    pub method_data: *mut libc::c_void,
+    pub free_func: Option<unsafe extern "C" fn(*mut libc::c_void) -> ()>,
+}
+
+pub struct GodotScriptClassBuilder<'l> {
+
+    is_tool: bool,
+
+    class_name: &'l str,
+    base_class_name: &'l str,
+
+    constructor: Option<GodotScriptConstructorType>,
+    destructor: Option<GodotScriptDestructorType>,
+
+    methods: Vec<GodotScriptMethod<'l>>,
+    properties: Vec<(String)>,
+    signals: Vec<(String)>,
+
+}
+
+impl<'l> GodotScriptClassBuilder<'l> {
+    pub fn new() -> Self {
+        GodotScriptClassBuilder {
+            is_tool: false,
+
+            class_name: "",
+            base_class_name: "",
+
+            constructor: None,
+            destructor: None,
+
+            methods: vec![],
+            properties: vec![],
+            signals: vec![],
+        }
+    }
+
+    pub fn set_tool(mut self, is_tool: bool) -> Self {
+        self.is_tool = is_tool;
+        self
+    }
+
+    pub fn set_class_name(mut self, name: &'l str) -> Self {
+        self.class_name = name;
+        self
+    }
+
+    pub fn set_base_class_name(mut self, name: &'l str) -> Self {
+        self.base_class_name = name;
+        self
+    }
+
+    pub fn add_method_advanced(mut self, method: GodotScriptMethod<'l>) -> Self {
+        self.methods.push(method);
+        self
+    }
+
+    pub fn add_method(mut self, name: &'l str, method: GodotScriptMethodType) -> Self {
+
+        let method = GodotScriptMethod {
+            name: name,
+            method_ptr: Some(method),
+			attributes: GodotScriptMethodAttributes {
```

Comment:
nit: indentation (looks like you added tabs rather than spaces).

Reply:
Oops, will fix!